### PR TITLE
feat(desktop): add configurable quit shortcut confirmation

### PR DESCRIPTION
**Pull request by juliusmarminge**
## Before

`confirmQuit` was a boolean. Disabled meant quit immediately. Enabled, which was the default, presented itself as hold to quit but accepted either of two gestures:

- Hold Cmd/Ctrl+Q for 1.2 seconds, then release.
- Press Cmd/Ctrl+Q again within 500 ms to quit immediately.

That second gesture was built into hold mode and was not configurable or explained in Settings. If someone canceled a hold and quickly tried again, the retry counted as the second press and quit the app. This is the behavior reported as a broken hold restart.

The application menu Quit action was separate and always quit immediately.

## After

`confirmQuit` is now an explicit mode:

- **Direct:** quit on the first Cmd/Ctrl+Q press.
- **Hold:** quit only after a completed 1.2-second hold. Canceling and pressing again starts a new hold, even inside the 500 ms window.
- **Double press:** quit when Cmd/Ctrl+Q is pressed twice within 500 ms. One press shows a "Press again to quit" hint until the window expires.

Hold remains the default. Existing stored booleans migrate as `true` to hold and `false` to direct. The application menu Quit action remains immediate.

The desktop shortcut handler, Settings selector, quit hint overlay, preload validation, and IPC contract now use the selected mode. Focused tests cover canceled holds, hold retries, quick and slow double presses, interrupted shortcuts, suppressed key-up events, and delayed settings reads.

## Rationale

The old boolean combined hold and double press into one setting. That made a deliberate hold retry indistinguishable from a double-press confirmation. Separating the gestures fixes that ambiguity and gives users the behavior they expect: immediate quit, hold only, or double press only.

## UI

General settings showing the quit shortcut mode selector

Implemented and reviewed with GPT-5.6 Sol through the Codex harness.

---

> [!NOTE]
> **Medium Risk**
> Changes how the desktop app quits from a global shortcut and alters the IPC contract for quit hints; mistakes could cause accidental quits or stuck overlays, though behavior is heavily tested.
> 
> **Overview**
> Desktop quit via **Cmd/Ctrl+Q** is no longer a single on/off “hold to quit” flag. **`confirmQuit`** becomes **`QuitConfirmationMode`**: **`direct`**, **`hold`**, or **`double-click`**, defaulting to **`hold`**. Stored booleans still decode (**`true` → hold**, **`false` → direct**) and re-encode as strings; **`ClientSettingsPatch`** no longer accepts boolean **`confirmQuit`**.
> 
> The main-process handler is renamed to **`makeQuitShortcutHandler`** and branches on the resolved mode: immediate quit, hold-then-release, or quit on two presses within **`QUIT_DOUBLE_PRESS_MS`**. Hint IPC moves from **`"down"` / `"up"`** strings to **`QuitShortcutHintEvent`** objects (including **`mode`** on press). Preload validates and forwards those shapes; **`QuitHoldOverlay`** shows hold vs “press again” copy with different dismiss timing.
> 
> General settings replace the quit switch with a **select**; search/reset copy refers to “quit shortcut” behavior. Contract and desktop tests cover migration, patch validation, and the expanded gesture edge cases (async mode reads, double-press windows, stale resolutions).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit edbbb40dbe9a3f0ad23727b0c515dc8d6c2cd49f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Add configurable quit shortcut modes (`direct`, `hold`, `double-click`) to desktop app
> - Replaces the boolean `confirmQuit` setting with a `QuitConfirmationMode` enum (`direct` | `hold` | `double-click`), defaulting to `hold`. Legacy boolean values are migrated on full decode but rejected in patches.
> - Refactors `makeQuitHoldHandler` into `makeQuitShortcutHandler` in [QuitHold.ts](https://github.com/pingdotgg/t3code/pull/9076/files#diff-c9e580dbc762f19a498b272467ed869973f2fb60fb1765ebbdf7b0d3c2883760), supporting all three modes with generation-tracked timers to prevent stale async resolutions.
> - Changes the `onQuitShortcut` IPC contract to emit structured `QuitShortcutHintEvent` objects (`{ state: 'down', mode }` or `{ state: 'up' }`) instead of plain strings, updated in [ipc.ts](https://github.com/pingdotgg/t3code/pull/9076/files#diff-b3653a8424fd66e30684757f240a41b389e093822f7c4139866bcae9e917a38a) and [preload.ts](https://github.com/pingdotgg/t3code/pull/9076/files#diff-eae0078d2bf4ac4beed2884f918b2816a97752917390f09980ba2cfad8c42cac).
> - Updates [QuitHoldOverlay.tsx](https://github.com/pingdotgg/t3code/pull/9076/files#diff-0d6c0b150ac2c5add6763dd61120bad7b1c955b39711c88119a1d021bfdb6998) to show mode-specific copy and adjust hint visibility timing per mode.
> - Updates the settings UI in [SettingsPanels.tsx](https://github.com/pingdotgg/t3code/pull/9076/files#diff-9dbbb4f86928d777550dead352b372341ccfb3f589782aa6d99b98f792c89d18) to use a Select control instead of a boolean toggle.
> - Risk: `onQuitShortcut` listeners must handle structured events instead of strings; any out-of-tree consumers of this IPC channel will break. The `confirmQuit` patch schema rejects legacy booleans, so clients sending `true`/`false` will fail validation.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized edbbb40.</sup>
> 
>

### CODE REVIEW

File: apps/desktop/src/window/QuitHold.ts [hunk 1]
```diff
@@ -132,27 +133,38 @@ export function makeQuitHoldHandler(
...
           return;
         }
-        armed = true;
+        if (
+          resolvedMode === "double-click" &&
+          previousPressAt !== 0 &&
+          now - previousPressAt <= QUIT_DOUBLE_PRESS_MS
+        ) {
+          quitNow();
+          return;
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/desktop/src/window/QuitHold.ts:161]: Fixed in 442645dc3. A physical key release no longer invalidates the pending mode read. Direct mode and a qualifying second press are honored after the read settles, while hold mode only arms if the press is still active. Added regression coverage for both releases beating their reads.

File: apps/web/src/components/QuitHoldOverlay.tsx [hunk 2]
```diff
@@ -21,26 +21,29 @@ export function QuitHoldOverlay() {
     const unsubscribe = subscribe((state) => {
       window.clearTimeout(hideTimer);
       if (state === "down") {
-        setVisible(true);
+        const mode = getClientSettings().confirmQuit;
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/web/src/components/QuitHoldOverlay.tsx]: Fixed in 442645dc3. The renderer no longer reads potentially unhydrated client settings to choose the message. It uses the mode resolved by the main process and carried in the hint event.

File: apps/desktop/src/window/DesktopWindow.ts [hunk 3]
```diff
@@ -551,12 +551,11 @@ export const make = Effect.gen(function* () {
...
-    // native menu sees it and only quit after the shortcut is held. The
-    // renderer shows the "Hold to Quit" hint via QUIT_SHORTCUT_CHANNEL.
-    const quitHoldHandler = makeQuitHoldHandler({
+    // Intercept the quit accelerator before the native menu sees it and apply
+    // the configured direct, hold, or double-press behavior.
+    const quitShortcutHandler = makeQuitShortcutHandler({
       platform: environment.platform,
-      isEnabled: () =>
+      getMode: () =>
         runPromise(
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/desktop/src/window/DesktopWindow.ts:559]: Fixed in 442645dc3. A physical key release no longer invalidates the pending mode read. Direct mode and a qualifying second press are honored after the read settles, while hold mode only arms if the press is still active. Added regression coverage for both releases beating their reads.

File: apps/web/src/components/QuitHoldOverlay.tsx [hunk 4]
```diff
@@ -21,26 +21,29 @@ export function QuitHoldOverlay() {
...
+        setVisibleMode(mode === "double-click" ? mode : "hold");
         return;
       }
-      hideTimer = window.setTimeout(() => setVisible(false), HIDE_AFTER_RELEASE_MS);
+      hideTimer = window.setTimeout(() => setVisibleMode(null), HIDE_AFTER_RELEASE_MS);
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/web/src/components/QuitHoldOverlay.tsx]: Fixed in 442645dc3. The main process now includes the resolved mode in each hint event, and the renderer hides double-click hints immediately on the up event while preserving the hold-mode linger.

[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/web/src/components/QuitHoldOverlay.tsx]: Fixed in 442645dc3. The main process now includes the resolved mode in each hint event, and the renderer hides double-click hints immediately on the up event while preserving the hold-mode linger.

File: apps/web/src/components/QuitHoldOverlay.tsx [hunk 5]
```diff
@@ -2,45 +2,52 @@ import { useEffect, useState } from "react";
...
+      if (hint.state === "down") {
+        pressedMode = hint.mode;
+        setVisibleMode(hint.mode);
         return;
       }
-      hideTimer = window.setTimeout(() => setVisible(false), HIDE_AFTER_RELEASE_MS);
+      if (pressedMode === "double-click") {
+        setVisibleMode(null);
+        return;
+      }
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/web/src/components/QuitHoldOverlay.tsx:31]: Fixed in c70fde3d5. Releasing Q or the modifier now preserves an active double-press hint and its watchdog. The main process emits the up event when the original 500 ms window expires or the attempt is cancelled. The deadline is measured from keydown, including when the settings read settles late. Added regression tests for both timing paths.

File: apps/desktop/src/window/QuitHold.ts [hunk 6]
```diff
@@ -107,10 +109,10 @@ export function makeQuitHoldHandler(
       // Any other key (or an extra modifier) pressed mid-hold breaks the
       // gesture; without this the hold timer keeps running through the
       // interruption and the next qualifying repeat would quit early. The
-      // interrupted press also stops counting toward a double tap — but only
+      // interrupted press also stops counting toward a double press, but only
       // here, not in release(), which runs mid-restart on an unseen-release
       // re-press and must not wipe that press's own tap timestamp.
-      if (holding && !input.isAutoRepeat) {
+      if ((holding || notified) && !input.isAutoRepeat) {
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/desktop/src/window/QuitHold.ts:119]: Fixed in edbbb40db. A Command or Control keydown is now recognized as the first half of the next full quit shortcut and no longer cancels the active double-press window. Added a regression that releases both keys, presses the modifier again, then completes the second shortcut.
